### PR TITLE
[Sub-ports] Added test case which validates tunneling over sub-ports

### PR DESCRIPTION
**Pull request by OleksandrKozodoi**
Signed-off-by: Oleksandr Kozodoi <oleksandrx.kozodoi@intel.com>

### Description of PR

Summary:

- Added new sub-ports tests: [test_tunneling_between_sub_ports](https://github.com/Azure/sonic-mgmt/blob/1af07ad0ccb9a9c5be112c1ab2469bef18eb0318/tests/sub_port_interfaces/test_sub_port_interfaces.py#L323-L356)
- Added fixture for setup tunnel configuration on the DUT and remove after tests: [apply_tunnel_table_to_dut](https://github.com/Azure/sonic-mgmt/blob/1af07ad0ccb9a9c5be112c1ab2469bef18eb0318/tests/sub_port_interfaces/conftest.py#L368-L401)
- Added new helpers methods
- Added template for setup tunnel configuration on the DUT: [tunnel_config.j2](https://github.com/Azure/sonic-mgmt/blob/1af07ad0ccb9a9c5be112c1ab2469bef18eb0318/tests/sub_port_interfaces/templates/tunnel_config.j2)
- Updated Test Plan
### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
Coverage of sub-ports feature by test cases to improve the quality of SONiC.
#### How did you do it?
Added new test cases to PyTest.
Implementation is based on the [Sub-ports design spec](https://github.com/Azure/SONiC/blob/master/doc/subport/sonic-sub-port-intf-hld.md)
#### How did you verify/test it?
`py.test --testbed=testbed-t0 --inventory=../ansible/lab --testbed_file=../ansible/testbed.csv --host-pattern=testbed-t0 -- module-path=../ansible/library sub_port_interfaces`
#### Any platform specific information?
SONiC Software Version: SONiC.master.162-dirty-20210331.125032
Distribution: Debian 10.9
Kernel: 4.19.0-12-2-amd64
Build commit: ecaf97d8
Build date: Wed Mar 31 13:09:26 UTC 2021
Platform: x86_64-arista_7170_32cd
HwSKU: Arista-7170-32CD-C32
ASIC: barefoot
#### Supported testbed topology if it's a new test case?
T0, T1

### CODE REVIEW

**Comment by OleksandrKozodoi** [comment]
@wangxin Could you please review?

**Comment by SavchukRomanLv** [comment]
@yxieca @lguohan @wangxin @neethajohn  can you please  review? we do have 2 more to open for finish testing activities

File: tests/sub_port_interfaces/conftest.py [hunk 1]
```diff
@@ -360,6 +365,42 @@ def apply_route_config_for_port(request, duthost, ptfhost, define_sub_ports_conf
...
+
+    new_sub_ports = apply_route_config['new_sub_ports']
+    sub_ports = apply_route_config['sub_ports']
+
+    for src_port in new_sub_ports:
+        tunnel_ip = sub_ports[src_port]['ip'].split('/')[0]
+        tunnel_addr_list.append(tunnel_ip)
+
+    tunnel_config_path = os.path.join(DUT_TMP_DIR, TUNNEL_TEMPLATE)
+    config_template = jinja2.Template(open(os.path.join(TEMPLATE_DIR, TUNNEL_TEMPLATE)).read())
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/conftest.py:400]: Done. Thanks!

[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/conftest.py]: Good idea. Thanks!

File: tests/sub_port_interfaces/sub_ports_helpers.py [hunk 2]
```diff
@@ -219,6 +234,72 @@ def generate_and_verify_icmp_traffic(duthost, ptfadapter, src_port, dst_port, ip
...
+        ip_dst: Destination IP address of DUT
+        pkt_action: Packet action (forwarded or drop)
+        tr_type: Type of traffic (TCP or UDP)
+        pktlen: packet length
+        ttl: Time to live
+        ip_tunnel: Tunnel IP address of DUT
+    """
+    # Get VLAN ID from name of sub-port
+    src_vlan_vid = int(src_port.split('.')[1])
+    dl_vlan_enable = True
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/sub_ports_helpers.py]: Done. Thanks!

File: tests/sub_port_interfaces/templates/tunnel_config.j2 [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    "TUNNEL": {
+{% for ip in tunnel_addr_list %}
+        "{{ "MuxTunnel" + loop.index|string }}": {
+            "dst_ip" : "{{ ip }}",
+            "ttl_mode": "pipe",
+            "ecn_mode": "copy_from_outer",
+            "dscp_mode": "pipe",
+            "tunnel_type": "IPINIP"
+        }{% if not loop.last %},
+{% endif %}{% endfor %}
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/templates/tunnel_config.j2]: Done.

File: tests/sub_port_interfaces/sub_ports_helpers.py [hunk 4]
```diff
@@ -79,6 +92,8 @@ def generate_and_verify_traffic(duthost, ptfadapter, src_port, dst_port, ip_src,
             generate_and_verify_tcp_udp_traffic(duthost, ptfadapter, src_port, dst_port, ip_src, ip_dst, tr_type, pktlen, ttl)
         elif 'ICMP' in tr_type:
             generate_and_verify_icmp_traffic(duthost, ptfadapter, src_port, dst_port, ip_src, ip_dst, pkt_action, tr_type, ttl)
+        elif 'decap' in tr_type:
+            generate_and_verify_decap_traffic(duthost, ptfadapter, src_port, dst_port, ip_src, ip_dst, pkt_action, tr_type, ttl, 104, ip_tunnel)
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/sub_ports_helpers.py]: Removed. Thanks!

File: tests/sub_port_interfaces/sub_ports_helpers.py [hunk 5]
```diff
@@ -219,6 +234,72 @@ def generate_and_verify_icmp_traffic(duthost, ptfadapter, src_port, dst_port, ip
...
+                        eth_src=src_mac,
+                        ip_src=ip_src,
+                        ip_dst=ip_dst,
+                        ip_tunnel=ip_tunnel,
+                        vlan_vid=src_vlan_vid,
+                        dl_vlan_enable=dl_vlan_enable,
+                        tr_type=tr_type,
+                        ttl=64)
+
+    # Build expected packet
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/sub_ports_helpers.py:272]: Good idea.

File: tests/sub_port_interfaces/sub_ports_helpers.py [hunk 6]
```diff
@@ -762,3 +847,14 @@ def remove_sub_port_from_ptf(ptfhost, sub_port, ip):
     """
     ptfhost.shell("ip address del {} dev {}".format(ip, sub_port))
     ptfhost.shell("ip link del {}".format(sub_port))
+
+
+def update_arp_table(duthost, ip):
+    """
+    Add entry to ARP table
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/sub_ports_helpers.py]: Done. Thanks!

File: tests/sub_port_interfaces/sub_ports_helpers.py [hunk 7]
```diff
@@ -762,3 +847,14 @@ def remove_sub_port_from_ptf(ptfhost, sub_port, ip):
     """
     ptfhost.shell("ip address del {} dev {}".format(ip, sub_port))
     ptfhost.shell("ip link del {}".format(sub_port))
+
+
+def update_arp_table(duthost, ip):
```
[the remark this answers is not in the document]
Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/sub_ports_helpers.py]: Good idea!

**Comment by OleksandrKozodoi** [comment]
@wendani Can you take a look? Thank you in advance.

File: tests/sub_port_interfaces/templates/tunnel_config.j2 [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    "TUNNEL": {
+{% for ip in tunnel_addr_list %}
+        "{{ "MuxTunnel" + loop.index|string }}": {
+            "dst_ip" : "{{ ip }}",
+            "ttl_mode": "pipe",
+            "ecn_mode": "copy_from_outer",
+            "dscp_mode": "pipe",
+            "tunnel_type": "IPINIP"
+        }{% if not loop.last %},{% endif %}
+{% endfor %}
```
Comment (wendani) [inline tests/sub_port_interfaces/templates/tunnel_config.j2:11]: An empty line should be needed between  10 and  11. Can you check `sonic-cfggen -t` output?

Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/templates/tunnel_config.j2:11]: Sure. Output:
```
$ sonic-cfggen -t tunnel_config.j2 
{
    "TUNNEL": {

        "MuxTunnel1": {
            "dst_ip" : "172.16.0.9",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        },

        "MuxTunnel2": {
            "dst_ip" : "172.16.4.13",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        }

    }
}
```
Do we need to add an empty line between 10 and 11?

Comment (wendani) [reply tests/sub_port_interfaces/templates/tunnel_config.j2:11]: Tried latest `sonic-cfggen` in vs docker. Vs docker built @sonic-buildimage master 1863e1fa2
```
# which sonic-cfggen 
/usr/local/bin/sonic-cfggen
# cat /usr/local/bin/sonic-cfggen       
#!/usr/bin/python3
```

#### Without empty line
```
$ cat tunnel_config.j2 
{
    "TUNNEL": {
{% for ip in tunnel_addr_list %}
        "{{ "MuxTunnel" + loop.index|string }}": {
            "dst_ip" : "{{ ip }}",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        }{% if not loop.last %},{% endif %}
{% endfor %}
    }
}
```
output
```
sonic-cfggen -t tunnel_config.j2 -a "{\"tunnel_addr_list\": [\"172.16.0.9\", \"172.16.4.13\"]}"
{
    "TUNNEL": {
        "MuxTunnel1": {
            "dst_ip" : "172.16.0.9",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        },        "MuxTunnel2": {
            "dst_ip" : "172.16.4.13",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        }    }
}
```

#### With empty line
```
$ cat tunnel_config.j2 
{
    "TUNNEL": {
{% for ip in tunnel_addr_list %}
        "{{ "MuxTunnel" + loop.index|string }}": {
            "dst_ip" : "{{ ip }}",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        }{% if not loop.last %},{% endif %}

{% endfor %}
    }
}
```
Pretty output
```
sonic-cfggen -t tunnel_config.j2 -a "{\"tunnel_addr_list\": [\"172.16.0.9\", \"172.16.4.13\"]}"
{
    "TUNNEL": {
        "MuxTunnel1": {
            "dst_ip" : "172.16.0.9",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        },
        "MuxTunnel2": {
            "dst_ip" : "172.16.4.13",
            "ttl_mode": "pipe",
            "ecn_mode": "copy_from_outer",
            "dscp_mode": "pipe",
            "tunnel_type": "IPINIP"
        }
    }
}
```

You can ignore if you do not see the above difference.

Comment (OleksandrKozodoi) [reply tests/sub_port_interfaces/templates/tunnel_config.j2:12]: It makes sense. Added an empty line to tunnel_config. Thanks for your help in investigating.

**Review by wendani — APPROVED** [review]

**Comment by OleksandrKozodoi** [comment]
@wangxin Can you take a look? Thank you!

**Comment by SavchukRomanLv** [comment]
@wangxin Can we merge this?

**Comment by SavchukRomanLv** [comment]
@lguohan @yxieca can this be merged? Thank you in advance.

**Review by neethajohn — APPROVED** [review]

**Comment by bingwang-ms** [comment]
A possible typo caused test failure of ```test_tunneling_between_sub_ports```, and I filed a PR to address it https://github.com/Azure/sonic-mgmt/pull/4450. Can you please help to review? Thanks

**Pull request merged**